File: mobile/lib/models/family.dart
/// Family and FamilyMember models.

class Family {
  final String id;
  final String name;
  final String createdBy;
  final List<FamilyMember> members;
  final DateTime createdAt;

  Family({
    required this.id,
    required this.name,
    required this.createdBy,
    this.members = const [],
    required this.createdAt,
  });

  factory Family.fromJson(Map<String, dynamic> json) {
    return Family(
      id: json['id'],
      name: json['name'],
      createdBy: json['created_by'],
      members: (json['members'] as List<dynamic>?)
              ?.map((m) => FamilyMember.fromJson(m))
              .toList() ??
          [],
      createdAt: DateTime.parse(json['created_at']),
    );
  }
}

class FamilyMember {
  final String id;
  final String userId;
  final String? userName;
  final String? userPhone;
  final String relationship;
  final String? nickname;
  final bool canEdit;
  final bool receivesSos;
  final bool receivesMissedAlerts;
  final DateTime createdAt;

  FamilyMember({
    required this.id,
    required this.userId,
    this.userName,
    this.userPhone,
    required this.relationship,
    this.nickname,
    this.canEdit = false,
    this.receivesSos = true,
    this.receivesMissedAlerts = true,
    required this.createdAt,
  });

  factory FamilyMember.fromJson(Map<String, dynamic> json) {
    return FamilyMember(
      id: json['id'],
      userId: json['user_id'],
      userName: json['user_name'],
      userPhone: json['user_phone'],
      relationship: json['relationship'],
      nickname: json['nickname'],
      canEdit: json['can_edit'] ?? false,
      receivesSos: json['receives_sos'] ?? true,
      receivesMissedAlerts: json['receives_missed_alerts'] ?? true,
      createdAt: DateTime.parse(json['created_at']),
    );
  }

  /// Display name: nickname or userName or relationship.
  String get displayName => nickname ?? userName ?? relationship;
}

File: backend/app/schemas/family.py
"""Family schemas."""

from pydantic import BaseModel
from typing import Optional, List
from uuid import UUID
from datetime import datetime


class FamilyCreate(BaseModel):
    name: str


class FamilyMemberAdd(BaseModel):
    phone: str  # Phone number of existing user OR invite
    relationship: str
    nickname: Optional[str] = None
    can_edit: bool = False
    receives_sos: bool = True
    receives_missed_alerts: bool = True


class FamilyMemberUpdate(BaseModel):
    relationship: Optional[str] = None
    nickname: Optional[str] = None
    can_edit: Optional[bool] = None
    receives_sos: Optional[bool] = None
    receives_missed_alerts: Optional[bool] = None


class FamilyMemberOut(BaseModel):
    id: UUID
    user_id: UUID
    user_name: Optional[str] = None
    user_phone: Optional[str] = None
    relationship: str
    nickname: Optional[str] = None
    can_edit: bool
    receives_sos: bool
    receives_missed_alerts: bool
    created_at: datetime

    model_config = {"from_attributes": True}


class FamilyOut(BaseModel):
    id: UUID
    name: str
    created_by: UUID
    members: List[FamilyMemberOut] = []
    created_at: datetime

    model_config = {"from_attributes": True}
